pkg/gadgets/trace/exec/tracer/tracer_test.go

```diff
@@ -0,0 +1,208 @@
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets/trace/exec/tracer"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets/trace/exec/types"
+	eventtypes "github.com/kinvolk/inspektor-gadget/pkg/types"
+)
+
+func TestCreateTracer(t *testing.T) {
+	utilstest.RequireRoot(t)
+
+	tracer, err := tracer.NewTracer(&tracer.Config{}, nil,
+		func(types.Event) {})
+	if err != nil {
+		t.Fatalf("Error creating tracer: %s", err)
+	}
+	if tracer == nil {
+		t.Fatal("Returned tracer was nil")
+	}
+	t.Cleanup(tracer.Stop)
+}
+
+func TestStopTracerIdempotent(t *testing.T) {
+	utilstest.RequireRoot(t)
+
+	tracer, err := tracer.NewTracer(&tracer.Config{}, nil,
+		func(types.Event) {})
+	if err != nil {
+		t.Fatalf("Error creating tracer: %s", err)
+	}
+	if tracer == nil {
+		t.Fatal("Returned tracer was nil")
+	}
+
+	// Check that a double stop doesn't cause issues
+	tracer.Stop()
+	tracer.Stop()
+}
+
+func createTracer(
+	t *testing.T, config *tracer.Config, callback func(types.Event),
+) *tracer.Tracer {
+	t.Helper()
+
+	tracer, err := tracer.NewTracer(config, nil, callback)
+	if err != nil {
+		t.Fatalf("Error creating tracer: %s", err)
+	}
+	return tracer
+}
+
+func TestTracer(t *testing.T) {
+	utilstest.RequireRoot(t)
+
+	manyArgs := []string{}
+	// 19 is DEFAULT_MAXARGS - 1 (-1 because args[0] is on the first position).
+	for i := 0; i < 19; i++ {
+		manyArgs = append(manyArgs, "/dev/null")
+	}
+
+	// Function to generate an event used most of the times.
+	// Returns pid of executed process.
+	generateEvent := func() (int, error) {
+		cmd := exec.Command("/bin/cat", "/dev/null")
+		if err := cmd.Run(); err != nil {
+			return 0, fmt.Errorf("running command: %w", err)
+		}
+
+		return cmd.Process.Pid, nil
+	}
+
+	tests := map[string]utilstest.Test[types.Event, int]{
+		"captures_all_events_without_configured_filters": {
+			GetTracer: func(info *utilstest.Info, eventCallback func(event types.Event)) utilstest.Tracer {
+				return createTracer(t, &tracer.Config{}, eventCallback)
+			},
+			GenerateEvent: generateEvent,
+			ValidateEvent: utilstest.ExpectAtLeastOneEvent(func(info *utilstest.Info, ret *int) *types.Event {
+				return &types.Event{
+					Event: eventtypes.Event{
+						Type: eventtypes.NORMAL,
+					},
+					Pid:       uint32(*ret),
+					Ppid:      uint32(info.Pid),
+					UID:       uint32(info.UID),
+					MountNsID: info.MountNsID,
+					Retval:    0,
+					Comm:      "cat",
+					Args:      []string{"/bin/cat", "/dev/null"},
+				}
+			}),
+		},
+		"filtering_no_event_should_be_captured": {
+			GetTracer: func(info *utilstest.Info, eventCallback func(event types.Event)) utilstest.Tracer {
+				config := &tracer.Config{
+					MountnsMap: utilstest.CreateMntNsFilterMap(t, 0),
+				}
+				return createTracer(t, config, eventCallback)
+			},
+			GenerateEvent: generateEvent,
+			ValidateEvent: utilstest.ExpectNoEvent[types.Event, int],
+		},
+		"filtering_one_event_should_be_captured": {
```

Comment:
This test and name seem to specific to me. If you test filtering by mount namespace here as far as I understand, shouldn't test generate an event in other mount namespace and verify that it does _not_ show up?

I'm not sure if you get what I mean here :smile: I mean if you test for at least one event, I understand you expect other events to show up, right? Then if you test the filtering, other events should also be occurring, but tracer is expected to not record them, right?

Reply:
> shouldn't test generate an event in other mount namespace and verify that it does not show up?

It's covered by the previous test case, an event is generated but it's not captured. 

I agree that it'll be nice to also generate a second event on a different mount namespace to be sure it's not captured, but unfortunately it'll require some rework of the current logic as we'd need to have a second runner per test case. It's something that we can do but I prefer to move with this and implement that later on.

Reply:
After the rework we did it was easier to implement this test. I did it for both open and exec tracers.